File: master_security_python/src/master_security/core/metrics.py
"""
Master Security Framework - Prometheus Metrics Engine
=====================================================

Native Prometheus metrics for all security operations.
Provides counters, histograms, and gauges for observability.

Metrics exposed:
    - msf_requests_total: Total requests processed
    - msf_threats_detected_total: Threats detected by category
    - msf_auth_attempts_total: Authentication attempts
    - msf_auth_failures_total: Authentication failures
    - msf_validation_duration_seconds: Validation latency
    - msf_crypto_operations_total: Cryptographic operations
    - msf_rate_limit_hits_total: Rate limit triggers
    - msf_active_sessions: Current active sessions
    - msf_policy_violations_total: Policy violations
    - msf_alerts_triggered_total: Security alerts

Usage:
    >>> from master_security.core import get_metrics
    >>> metrics = get_metrics()
    >>> metrics.inc_counter("auth_attempts", user_type="premium")
"""

from __future__ import annotations

import threading
from typing import Any, Optional

from prometheus_client import Counter, Histogram, Gauge, CollectorRegistry

from master_security.core.logger import get_logger

logger = get_logger("msf.metrics")


class MetricsRegistry:
    """
    Prometheus metrics registry for MSF.

    Thread-safe metrics collection with automatic labeling.

    Attributes:
        registry: Prometheus CollectorRegistry
        prefix: Metric name prefix

    Example:
        >>> registry = MetricsRegistry()
        >>> registry.inc_counter("requests", method="GET")
    """

    def __init__(
        self,
        prefix: str = "msf",
        registry: Optional[CollectorRegistry] = None,
    ) -> None:
        self.prefix = prefix
        self.registry = registry or CollectorRegistry()
        self._counters: dict[str, Counter] = {}
        self._histograms: dict[str, Histogram] = {}
        self._gauges: dict[str, Gauge] = {}
        self._lock = threading.Lock()

    def _counter_name(self, name: str) -> str:
        return f"{self.prefix}_{name}_total"

    def inc_counter(
        self,
        name: str,
        value: float = 1.0,
        labels: Optional[dict[str, str]] = None,
    ) -> None:
        """
        Increment a counter metric.

        Args:
            name: Metric name (without prefix)
            value: Amount to increment
            labels: Optional label key-value pairs
        """
        full_name = self._counter_name(name)
        if full_name not in self._counters:
            with self._lock:
                if full_name not in self._counters:
                    self._counters[full_name] = Counter(
                        full_name,
                        f"MSF {name} total",
                        labelnames=sorted(labels.keys()) if labels else [],
                        registry=self.registry,
                    )
        if labels:
            self._counters[full_name].labels(**labels).inc(value)
        else:
            self._counters[full_name].inc(value)

    def observe_histogram(
        self,
        name: str,
        value: float,
        labels: Optional[dict[str, str]] = None,
    ) -> None:
        """
        Record a histogram observation.

        Args:
            name: Metric name
            value: Observed value
            labels: Optional labels
        """
        full_name = f"{self.prefix}_{name}_seconds"
        if full_name not in self._histograms:
            with self._lock:
                if full_name not in self._histograms:
                    self._histograms[full_name] = Histogram(
                        full_name,
                        f"MSF {name} duration",
                        labelnames=sorted(labels.keys()) if labels else [],
                        buckets=(0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0),
                        registry=self.registry,
                    )
        if labels:
            self._histograms[full_name].labels(**labels).observe(value)
        else:
            self._histograms[full_name].observe(value)

    def set_gauge(
        self,
        name: str,
        value: float,
        labels: Optional[dict[str, str]] = None,
    ) -> None:
        """
        Set a gauge metric.

        Args:
            name: Metric name
            value: Gauge value
            labels: Optional labels
        """
        full_name = f"{self.prefix}_{name}"
        if full_name not in self._gauges:
            with self._lock:
                if full_name not in self._gauges:
                    self._gauges[full_name] = Gauge(
                        full_name,
                        f"MSF {name}",
                        labelnames=sorted(labels.keys()) if labels else [],
                        registry=self.registry,
                    )
        if labels:
            self._gauges[full_name].labels(**labels).set(value)
        else:
            self._gauges[full_name].set(value)

    def inc_gauge(
        self,
        name: str,
        value: float = 1.0,
        labels: Optional[dict[str, str]] = None,
    ) -> None:
        """Increment a gauge metric."""
        full_name = f"{self.prefix}_{name}"
        if full_name not in self._gauges:
            self.set_gauge(name, 0, labels)
        if labels:
            self._gauges[full_name].labels(**labels).inc(value)
        else:
            self._gauges[full_name].inc(value)

    def dec_gauge(
        self,
        name: str,
        value: float = 1.0,
        labels: Optional[dict[str, str]] = None,
    ) -> None:
        """Decrement a gauge metric."""
        full_name = f"{self.prefix}_{name}"
        if full_name in self._gauges:
            if labels:
                self._gauges[full_name].labels(**labels).dec(value)
            else:
                self._gauges[full_name].dec(value)

    def get_metrics_text(self) -> str:
        """Get all metrics as Prometheus text format."""
        from prometheus_client import generate_latest
        return generate_latest(self.registry).decode("utf-8")


_global_metrics: MetricsRegistry | None = None
_metrics_lock = threading.Lock()


def get_metrics() -> MetricsRegistry:
    """
    Get the global metrics registry.

    Returns:
        Global MetricsRegistry instance.
    """
    global _global_metrics
    if _global_metrics is None:
        with _metrics_lock:
            if _global_metrics is None:
                _global_metrics = MetricsRegistry()
                logger.info("msf.metrics.initialized")
    return _global_metrics

File: packages/core/src/core/metrics.ts
/**
 * Metrics registry for collecting and exposing application metrics
 * @module core/metrics
 */

/**
 * Metric type enumeration
 */
export enum MetricType {
  /** Monotonically increasing counter */
  COUNTER = 'counter',
  /** Distribution of values */
  HISTOGRAM = 'histogram',
  /** Point-in-time value */
  GAUGE = 'gauge',
}

/**
 * Counter metric for monotonically increasing values
 */
export interface CounterMetric {
  /** Metric name */
  name: string;
  /** Metric description */
  description: string;
  /** Current value */
  value: number;
  /** Metric labels */
  labels: Record<string, string>;
}

/**
 * Histogram metric for value distributions
 */
export interface HistogramMetric {
  /** Metric name */
  name: string;
  /** Metric description */
  description: string;
  /** Bucket boundaries */
  buckets: number[];
  /** Bucket counts */
  bucketCounts: number[];
  /** Sum of all observed values */
  sum: number;
  /** Total count of observations */
  count: number;
  /** Metric labels */
  labels: Record<string, string>;
}

/**
 * Gauge metric for point-in-time values
 */
export interface GaugeMetric {
  /** Metric name */
  name: string;
  /** Metric description */
  description: string;
  /** Current value */
  value: number;
  /** Metric labels */
  labels: Record<string, string>;
}

/**
 * Metrics registry for managing counters, histograms, and gauges
 */
export class MetricsRegistry {
  /** Counter metrics storage */
  private counters: Map<string, CounterMetric>;
  /** Histogram metrics storage */
  private histograms: Map<string, HistogramMetric>;
  /** Gauge metrics storage */
  private gauges: Map<string, GaugeMetric>;

  /**
   * Create a new MetricsRegistry
   */
  constructor() {
    this.counters = new Map();
    this.histograms = new Map();
    this.gauges = new Map();
  }

  /**
   * Register a new counter metric
   * @param name - Metric name
   * @param description - Metric description
   * @param labels - Optional labels
   * @example
   * ```typescript
   * registry.registerCounter('http_requests_total', 'Total HTTP requests', { method: 'GET' });
   * ```
   */
  registerCounter(name: string, description: string, labels: Record<string, string> = {}): void {
    this.counters.set(name, { name, description, value: 0, labels });
  }

  /**
   * Register a new histogram metric
   * @param name - Metric name
   * @param description - Metric description
   * @param buckets - Bucket boundaries
   * @param labels - Optional labels
   * @example
   * ```typescript
   * registry.registerHistogram('http_request_duration_ms', 'Request duration', [5, 10, 25, 50, 100, 250, 500, 1000]);
   * ```
   */
  registerHistogram(
    name: string,
    description: string,
    buckets: number[] = [5, 10, 25, 50, 100, 250, 500, 1000],
    labels: Record<string, string> = {}
  ): void {
    this.histograms.set(name, {
      name,
      description,
      buckets,
      bucketCounts: new Array(buckets.length + 1).fill(0),
      sum: 0,
      count: 0,
      labels,
    });
  }

  /**
   * Register a new gauge metric
   * @param name - Metric name
   * @param description - Metric description
   * @param labels - Optional labels
   * @example
   * ```typescript
   * registry.registerGauge('active_connections', 'Current active connections');
   * ```
   */
  registerGauge(name: string, description: string, labels: Record<string, string> = {}): void {
    this.gauges.set(name, { name, description, value: 0, labels });
  }

  /**
   * Increment a counter by the specified amount
   * @param name - Counter name
   * @param value - Amount to increment (default 1)
   * @example
   * ```typescript
   * registry.incCounter('http_requests_total');
   * registry.incCounter('http_requests_total', 5);
   * ```
   */
  incCounter(name: string, value: number = 1): void {
    const counter = this.counters.get(name);
    if (!counter) {
      this.registerCounter(name, name);
      this.counters.get(name)!.value += value;
      return;
    }
    counter.value += value;
  }

  /**
   * Observe a value in a histogram
   * @param name - Histogram name
   * @param value - Value to observe
   * @example
   * ```typescript
   * registry.observeHistogram('http_request_duration_ms', 42);
   * ```
   */
  observeHistogram(name: string, value: number): void {
    const histogram = this.histograms.get(name);
    if (!histogram) {
      this.registerHistogram(name, name);
      this.observeHistogram(name, value);
      return;
    }
    histogram.sum += value;
    histogram.count += 1;
    for (let i = 0; i < histogram.buckets.length; i++) {
      if (value <= histogram.buckets[i]) {
        histogram.bucketCounts[i] += 1;
        return;
      }
    }
    histogram.bucketCounts[histogram.bucketCounts.length - 1] += 1;
  }

  /**
   * Set a gauge to an absolute value
   * @param name - Gauge name
   * @param value - Value to set
   * @example
   * ```typescript
   * registry.setGauge('active_connections', 42);
   * ```
   */
  setGauge(name: string, value: number): void {
    const gauge = this.gauges.get(name);
    if (!gauge) {
      this.registerGauge(name, name);
      this.gauges.get(name)!.value = value;
      return;
    }
    gauge.value = value;
  }

  /** Alias for setGauge */
  gauge(name: string, value: number, labels?: Record<string, string>): void {
    if (labels) {
      this.registerGauge(name, name, labels);
    }
    this.setGauge(name, value);
  }

  /**
   * Increment a gauge by the specified amount
   * @param name - Gauge name
   * @param value - Amount to increment (default 1)
   * @example
   * ```typescript
   * registry.incGauge('active_connections');
   * ```
   */
  incGauge(name: string, value: number = 1): void {
    const gauge = this.gauges.get(name);
    if (!gauge) {
      this.registerGauge(name, name);
      this.gauges.get(name)!.value += value;
      return;
    }
    gauge.value += value;
  }

  /**
   * Decrement a gauge by the specified amount
   * @param name - Gauge name
   * @param value - Amount to decrement (default 1)
   * @example
   * ```typescript
   * registry.decGauge('active_connections');
   * ```
   */
  decGauge(name: string, value: number = 1): void {
    const gauge = this.gauges.get(name);
    if (!gauge) {
      this.registerGauge(name, name);
      this.gauges.get(name)!.value -= value;
      return;
    }
    gauge.value -= value;
  }

  /**
   * Get all metrics in a structured format
   * @returns Object containing all metric types
   * @example
   * ```typescript
   * const metrics = registry.getMetrics();
   * console.log(metrics.counters, metrics.histograms, metrics.gauges);
   * ```
   */
  getMetrics(): {
    counters: Map<string, CounterMetric>;
    histograms: Map<string, HistogramMetric>;
    gauges: Map<string, GaugeMetric>;
  } {
    return {
      counters: new Map(this.counters),
      histograms: new Map(this.histograms),
      gauges: new Map(this.gauges),
    };
  }

  /**
   * Start a timer for measuring duration
   * @param name - Histogram metric name
   * @returns Object with stop() method
   */
  startTimer(name: string): { stop(): void } {
    const start = Date.now();
    return {
      stop: () => {
        const duration = Date.now() - start;
        this.observeHistogram(name, duration);
      },
    };
  }

  /**
   * Clear all metrics
   */
  clear(): void {
    this.counters.clear();
    this.histograms.clear();
    this.gauges.clear();
  }
}

let _metrics: MetricsRegistry | null = null;

/**
 * Get the global metrics registry singleton
 * @returns MetricsRegistry instance
 * @example
 * ```typescript
   const metrics = getMetrics();
   * metrics.incCounter('requests_total');
   * ```
   */
export function getMetrics(): MetricsRegistry {
  if (!_metrics) {
    _metrics = new MetricsRegistry();
  }
  return _metrics;
}
